src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -1634,5 +1634,41 @@ internal bool IsSimpleProgram
                 return this.declaration.Declarations.Any(d => d.IsSimpleProgram);
             }
         }
+
+        protected override void AfterMembersCompletedChecks(BindingDiagnosticBag diagnostics)
+        {
+            base.AfterMembersCompletedChecks(diagnostics);
+
+            // We need to give warnings if Obsolete is applied to any required members and there are constructors where a user would be forced to set that member,
+            // unless:
+            //  1. We're in an obsolete context ourselves, or
+            //  2. All constructors of this type are obsolete or attributed with SetsRequiredMembersAttribute
+            // We don't warn for required members from base types, as the user either has a warning that they're depending on an obsolete constructor/type already,
+            // or the original author ignored this warning.
+
+            // Obsolete states should have already been calculated by this point in the pipeline.
+            Debug.Assert(ObsoleteKind != ObsoleteAttributeKind.Uninitialized);
+            Debug.Assert(GetMembers().All(m => m.ObsoleteKind != ObsoleteAttributeKind.Uninitialized));
+
+            if (ObsoleteKind != ObsoleteAttributeKind.None
+                || GetMembers().All(m => m is not MethodSymbol { MethodKind: MethodKind.Constructor, ObsoleteKind: ObsoleteAttributeKind.None }
+                                         || !((MethodSymbol)m).ShouldCheckRequiredMembers()))
```

Comment:
might be a little easier to read if the pattern on the previous line introduced a 'method' local instead of casting here.